### PR TITLE
[BugFix](functions) fix multi_search_all_positions

### PR DESCRIPTION
**Pull request by zclllyybb**
# Proposed changes

before:
```
+---------------+----------------------------------+-----------------------------------------------+
| content       | mode                             | multi_search_all_positions(`content`, `mode`) |
+---------------+----------------------------------+-----------------------------------------------+
| Hello, World! | ['hello', 'world']               | [0, 0]                                        |
| Hello, World! | ['hello', 'world', 'Hello', '!'] | [0, 0, 0, 0]                                  |
| hello, world! | ['Hello', 'world']               | [0, 0]                                        |
| hello, world! | ['hello', 'world', 'Hello', '!'] | [0, 0, 0, 0]                                  |
| HHHHW!        | ['H', 'HHHH', 'HW', 'WH']        | [0, 0, 0, 0]                                  |
+---------------+----------------------------------+-----------------------------------------------+
```

after:
```
+---------------+----------------------------------+-----------------------------------------------+
| content       | mode                             | multi_search_all_positions(`content`, `mode`) |
+---------------+----------------------------------+-----------------------------------------------+
| Hello, World! | ['hello', 'world']               | [0, 0]                                        |
| Hello, World! | ['hello', 'world', 'Hello', '!'] | [0, 0, 1, 13]                                 |
| hello, world! | ['Hello', 'world']               | [0, 8]                                        |
| hello, world! | ['hello', 'world', 'Hello', '!'] | [1, 8, 0, 13]                                 |
| HHHHW!        | ['H', 'HHHH', 'HW', 'WH']        | [1, 1, 4, 0]                                  |
+---------------+----------------------------------+-----------------------------------------------+
```

## Problem summary

Completely rewrite the function. now it search for every needle's **first** occur in haystack.

## Checklist(Required)

* [ ] Does it affect the original behavior
* [x] Has unit tests been added
* [x] Has document been added or modified
* [ ] Does it need to update dependencies
* [x] Is this PR support rollback (If NO, please explain WHY)

### CODE REVIEW

File: be/src/vec/functions/functions_multi_string_position.cpp [hunk 1]
```diff
@@ -103,8 +105,19 @@ class FunctionMultiStringPosition : public IFunction {
     }
 };
 
-template <typename Impl>
 struct FunctionMultiSearchAllPositionsImpl {
+private:
+    using SingleSearcher = ASCIICaseSensitiveStringSearcher;
+    static SingleSearcher create_multi_searcher(const StringRef& needle) {
```
Comment (HappenLee) [inline be/src/vec/functions/functions_multi_string_position.cpp:111]: only use in `FunctionMultiSearchAllPositionsImpl ` do not need static

File: be/src/vec/functions/functions_multi_string_position.cpp [hunk 2]
```diff
@@ -113,38 +126,50 @@ struct FunctionMultiSearchAllPositionsImpl {
...
+        for (const auto& needle : needles_arr) {
+            needles.emplace_back(needle.get<StringRef>());
+        }
 
-        const size_t haystack_size = haystack_offsets.size();
         const size_t needles_size = needles.size();
+        std::vector<SingleSearcher> searchers;
+        searchers.reserve(needles_size);
+        for (auto needle : needles) {
+            searchers.emplace_back(needle.data, needle.size);
```
Comment (HappenLee) [inline be/src/vec/functions/functions_multi_string_position.cpp:146]: why do the work in up for-loop ？

**Comment by zclllyybb** [comment]
run buildall

**Comment by zclllyybb** [comment]
run p1

**Review by HappenLee — APPROVED** [review]
LGTM

**Pull request merged**